### ISSUE TITLE
Building a Huberman Lab Podcast Search Engine Using Hugging Face

### ISSUE THREAD

**Issue by Stepheni12** [Stepheni12/iSteph#1]
---
title: Building a Huberman Lab Podcast Search Engine Using Hugging Face
description: I built a "Space" in Hugging Face that allows you to search for topics across Huberman Lab podcast episodes to more easily find relevant episodes/segments for what you are looking for.
slug: building-a-hugging-face-space
tags: whisper, hugging face, gradio, podcasts
category: blog
date: 2024-01-24
---

I wanted to do a project to get more experience with Hugging Face and Gradio and figured this would be a fun project to go after. This isn't meant to be a tutorial, but I'm going to go through each of the steps I went through to build this Hugging Face Space.

TL;DR I made a Hugging Face [Space](https://huggingface.co/spaces/Stepheni12/Huberman_Topic_Search) that lets you search through Huberman Lab podcasts let me know what you think!

### General Idea
The idea behind this Space is that I am going to separate every podcast transcript into sections and generate embeddings for all of these individual sections. Then when you want to search for a specific topic you can type it in and it will then generate an embedding for your query and return the "n" closest sections to what you searched for. In this case "closeness" is ranked by measuring the distance between the vector representing the embedding for your search query and the vector of each of the individual sections from all the podcast episodes. Ultimately, you'll end up with a list of excerpts from episodes most closely related to the topic you are searching for. Check [this](https://www.turing.com/kb/guide-on-word-embeddings-in-nlp) out if you need more info on embeddings. These are the general steps I followed:

1. Download mp3s of every podcast episode
2. Use whisper through HF to generate transcripts of each mp3
3. Separate every transcript into sections and generate text embeddings for each section
4. Store embeddings in a vector database to easily query searches against
5. Build the Gradio app
5. Publish as a HF Space

### Getting the data
To acquire the data I went on the Huberman Lab [website](https://www.hubermanlab.com/podcast) and found the XML file representing the entire RSS feed. Within this XML I was able to pull the mp3 files of each podcast episode as well as the title of that episode.

![xml_pic](https://github.com/Stepheni12/iSteph/assets/26886046/8a8fe8bd-aeac-444b-8e36-ca5726d1973c)

Once I scraped all the mp3 links I created a Python script to go through and download each of those mp3s. This took quite a while, I think I may have just let it run overnight.

### Generating transcripts
Now that I had all the mp3 files for each of the episodes I needed to come up with an efficient way of generating the transcripts. My issue is that I only have my laptop which is nearly 10 years old at this point, (still going strong though!) no GPU or anything. I had a few ideas though.
- Use the free tier of Google Colab GPU to slowly chip away at generating the transcripts
- Use a Kaggle GPU notebook to do the same thing
- Use my brother's gaming desktop to generate the transcripts while he is at work

I ended up going with the 3rd option, far less tedious. I did consider using the tiny whisper model as my laptop was able to run that (it did a decent job transcribing Kanye's "I Love Kanye" if you were curious) but it would've taken multiple overnight runs and on top of that the vocabulary of the Huberman Lab podcast episodes gets quite complex with all the medical terms. Ultimately I ended up using the medium whisper model which is roughly 20x bigger than the tiny one. I generated all the transcripts over a few days while my brother was at work.

### Storing the embeddings in a vector database
For storing all this transcript information I utilized Chroma DB. I've never used a vector database before so I watched a couple of tutorials and read the quickstart and usage documentation a few times before figuring out the best way to apply the examples to my dataset. Chroma DB is open source and super simple to use, and it being so easy to set up a persistent DB was pretty appealing, although to be honest I didn't look into many other options.
> Note: Chroma is cool because you can input the raw text to be added to the database and it will automatically use HF to generate the embeddings for you. The default embedding is [this](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2), but beware because this default will automatically truncate the raw text you input down to 256 words, and in the case of these podcast episodes this eliminates 99.99% percent of the content. At first, I didn't realize this, so I had to circle back and convert the full text into sections of 254 words each (I chose 254 just to have a little buffer).

### Building the Gradio app
This was also my first time using Gradio, and for something simple like what I was creating in this project, it is super easy to use. I read through the quickstart and certain parts of the documentation a couple of times and this was enough to build the app.

![gradio_pic](https://github.com/Stepheni12/iSteph/assets/26886046/0eb4dd4f-0cd5-402f-a440-bd8bd6b61625)

This was essentially all the code to build the app and you just pass it the function you want the app to execute, so it could be something as simple as a Python function that adds two numbers or a function that runs the prediction of data through a trained ML model you have, it's worth checking out if you haven't already.

### Publishing it as a HF Space
Publishing the app as a Space is pretty simple to do because it operates just like a GitHub repo. Check out the [Space](https://huggingface.co/spaces/Stepheni12/Huberman_Topic_Search) and let me know what you think. It's not fully finished yet and could use some design improvements, but I think it is useful already!

> Feel free to skip this last part, it's mostly me just letting off some anger about my bad internet connection. I am leaving it in though in case it happens to help someone. Long story short, pushing to a remote repo might be failing because of a bad internet connection.

For me, this process of publishing the Space was quite frustrating. Not at all because of HF(well maybe a tad), but because I have a pretty bad internet connection in certain parts of the house. Every HF Space is a GitHub repo and you just clone the repo and then push your gradio or streamlit app (or whatever app) and any additional code or data. In my case, I was pushing my app.py file for gradio and the chroma database I created that had all the embedding information. The database isn't huge ~80MB, but because of the size, you have to use git lfs because it's considered a large file. I kept running into this issue where every time I tried to push to the repo I would get various errors and it would fail. I initially thought it was an authentication issue so I went down that rabbit hole. Then I thought it was because I was doing something wrong with git lfs so I ended up cloning and deleting the repo like 10 times. Finally, I took my laptop out from where I normally work and walked over to the room with our router in it and everything worked fine. It would've been a lot easier if it had just pointed out it was a connection issue because it wasn't at all obvious, but maybe that's hard for Hugging Face or github to realize, I don't know.

#### Note for myself
First blog post in the books! Along with becoming a better ML Engineer I also want to improve my writing and I'm going to work on that using Ben Franklin's [technique](https://observer.com/2016/12/the-benjamin-franklin-method-how-to-actually-be-a-better-writer/). I wanted to get this out publicly to hold myself accountable. If you are seeing this in the future and it is clear my writing hasn't improved then link me to this section and I will buy you a coffee!
